tests/test_utils.py

```diff
@@ -19,3 +31,145 @@ def test_route_params():
+    assert res["optimizer"][2][1].got["kwarg"] == 4
+    assert isinstance(res["optimizer"][3], dict)
+    assert set(res["optimizer"][3]["foo1"].got.keys()) == {"kwarg"}
+    assert res["optimizer"][3]["foo1"].got["kwarg"] == 5
+    assert set(res["optimizer"][3]["foo2"].got.keys()) == set()
+    assert res["optimizer"][4] == "foostring"
+    assert res["optimizer"][5] is instance
+    assert set(res["optimizer"][5].got.keys()) == {"kwarg"}
+    assert res["optimizer"][5].got["kwarg"] == 7
+    assert isinstance(res["optimizer"][6], tuple)
+    assert isinstance(res["optimizer"][6][0], Foo)
+    assert set(res["optimizer"][6][0].got.keys()) == set()
+
+
+def test_compile_with_params_dependency():
+    """Test the `compile_with_params` function's
+    ability to compile the params themselves.
+    """
+
+    class Foo:
+        got = dict()
+
+        def __init__(self, **kwargs):
+            self.got = kwargs
+
+    class Bar:
+        got = dict()
+
+        def __init__(self, **kwargs):
+            self.got = kwargs
+
+    res = compile_with_params(
+        items={"optimizer": [Foo, [Foo,]],},  # 0  # 1
+        params={
+            "optimizer__0__bar": Bar,
+            "optimizer__0__bar_class": Bar,
+            "optimizer__0__bar__bar_kwarg": 10,
+            "optimizer__1__0__bar": Bar,
+            "optimizer__1__0__bar__bar_kwarg": 11,
+        },
+    )
+    # Checks
+    assert isinstance(res["optimizer"][0].got["bar"], Bar)
+    assert isinstance(res["optimizer"][0].got["bar_class"], Bar)
+    assert set(res["optimizer"][0].got["bar_class"].got.keys()) == set()
+    assert set(res["optimizer"][0].got["bar"].got.keys()) == {"bar_kwarg"}
+    assert res["optimizer"][0].got["bar"].got["bar_kwarg"] == 10
+    assert isinstance(res["optimizer"][1][0].got["bar"], Bar)
+    assert set(res["optimizer"][1][0].got["bar"].got.keys()) == {"bar_kwarg"}
+    assert res["optimizer"][1][0].got["bar"].got["bar_kwarg"] == 11
+
+
+def test_compile_with_params_params_to_uncompilable():
+    """Test that a warning is raised if parameters are routed
+    to a non-class.
+    """
+    with pytest.warns(
+        UserWarning, match="SciKeras does not know how to compile"
+    ):
+        compile_with_params(
+            items={"test": "not_a_class"}, params={"test__kwarg": None},
+        )
+
+
+@pytest.mark.parametrize(
+    "items,expected,item_type",
+    [
+        ("binary_crossentropy", losses_module.binary_crossentropy, "loss"),
+        ("BinaryCrossentropy", losses_module.BinaryCrossentropy, "loss"),
+        ("SGD", optimizers_module.SGD, "optimizer"),
+        ("binary_accuracy", metrics_module.binary_accuracy, "metrics"),
+        ("BinaryAccuracy", metrics_module.BinaryAccuracy, "metrics"),
+        ("Unknown", "Unknown", "loss"),
+        ("Unknown", "Unknown", "optimizer"),
+        ("Unknown", "Unknown", "metrics"),
+    ],
+)
+def test_class_from_strings(items, expected, item_type):
```

Comment:
The test you are suggesting exists: https://github.com/adriangb/scikeras/blob/880f16c7611b941e3d7151ca8476ea1a2f2af9e8/tests/test_compile_kwargs.py#L143

This test is specifically for `_class_from_strings`. I do not see anything wrong with testing this method within the testing module designated to test it's module.

Reply:
> I also think that if the value can't be obtained from tf.keras.losses.get an error should be raised.

I generally feel that as a wrapper library, anything that ultimately causes an error in TF we should leave for TF to handle. Unless the error is very cryptic, in which case we bend this rule a bit and step in to help users out a bit, TF errors are not great and I don't want to be _cruel_.

Reply:
> I do not see anything wrong with testing this method within the testing module designated to test it's module.

Why would you test a private method, one that users will not use? If that test fails, what consequences do the users and developers experience?

Reply:
I tried to reply via email. It ended up in https://github.com/adriangb/scikeras/pull/66#issuecomment-696427413

Reply:
Great. But that's not why I write tests. I write test during development to make sure my library works as expected for my users. I write code to help with development all the time, but I don't encode that behavior in my library's tests.

I write tests to ensure my software works as expected for *users*, not for developers. So why doesn't your other tests fail given improper input? When I ran `test_loss_iterable` with `loss in ["junk", Foo, Foo()]`, it raised an error each time (and succeeded for other values of `loss`). Doesn't that test provide complete test coverage of this private method? Private methods are an implementation detail, the reason 100% test coverage is so difficult.

More reasons testing private functions is a bad practice: https://softwareengineering.stackexchange.com/questions/380287/why-is-unit-testing-private-methods-considered-as-bad-practice

Reply:
I personally think there are also valid reasons to test private methods, but I am here to learn so thank you for the links and info. I have seen many examples of notable libraries testing private methods. The first that comes to mind is sklearn: [example method](https://github.com/scikit-learn/scikit-learn/blob/74b632a1aa880a9c7f855599d16603626edf97fa/sklearn/utils/_encode.py#L52), [example test](https://github.com/scikit-learn/scikit-learn/blob/74b632a1aa880a9c7f855599d16603626edf97fa/sklearn/utils/tests/test_encode.py#L19). Are they then violating best practices or am I missing something here?

Do you feel that we could structure the package differently or name things differently to avoid issues like this?

Regarding your proposed test, I am happy to implement it, it is a good first step towards not needing this test in the first place.

Reply:
That's a good example. If you look at the original PR, https://github.com/scikit-learn/scikit-learn/pull/10209, you'll only see one test where the test `_encode` (the same tests you pointed out, `test_encode_util`). In general, the API behind private functions is not guaranteed to be stable:

> Private API, including functions, modules and methods beginning `_` are not assured to be stable.
>
> —https://scikit-learn.org/stable/glossary.html#term-api

Meaning that the tests are not guaranteed. More generally, I will consider testing private functions if any of the following applies:

* The features of the function are explicitly mentioned in the documentation.
* The function provides a core feature (e.g, "I'm using this library because of {core feature}").

I will consider testing private functions if they meet either of these criteria, but it's still not guaranteed. For example, in Dask-ML I wrote some private code that made a certain task faster. I mentioned this feature in the docs, [a blog post] and in two talks/papers. After some convincing, I wrote a test to make sure that speed benefit worked as expected.

[a blog post]:https://blog.dask.org/2019/09/30/dask-hyperparam-opt#dask-opportunities

Reply:
This whole conversation might be pointless: why not make `_class_from_string` a dictionary? 

``` python
def _get(name: str, fn : callable = tf.keras.losses.get):
    if name in ["get", "serialize", "deserialize"] or name[0] == "_":
        return None
    try:
        return fn(name):
    except ValueError:
        return None

_losses = {k: _get(k) for k in dir(tf.keras.losses) if _get(k)}
_optimizers = {k.lower(): _get(k, fn=tf.keras.optimizers.get) for k in dir(tf.keras.losses) _get(k, fn=tf.keras.optimizers.get)}
```

Dictionaries are the appropriate data structure: they can server as a map between strings and classes.

Reply:
This would replace the portion of `_class_from_string` that calls `tf.keras.{losses, optimizers, metrics}.get` right? Will we still keep the portion that recurses through iterables and dicts.

Reply:
Here's the implementation I'm thinking of:

``` python
###########  file: scikeras/utils.py
def _get(name: str, fn: callable = tf.keras.losses.get):
    if name in ["get", "serialize", "deserialize", "Loss"] or name[0] == "_":
        return None
    try:
        return fn(name)
    except ValueError:
        return None

losses = {k: _get(k) for k in dir(tf.keras.losses) if _get(k)}
optimizers = {k.lower(): _get(k, fn=tf.keras.optimizers.get) for k in dir(tf.keras.optimizers) if _get(k, fn=tf.keras.optimizers.get)}

########### file: tests/test_utils.py
from scikeras.utils import losses, optimizers

def _build_keras_model(loss: Union[str, Loss, function] = "mse", optimizer: Union[str, Optimizer] = "sgd"):
    if isinstance(loss, str):
        loss = losses[loss]
    if isinstance(optimizer, str):
        optimizer = optimizers[optimizer]
    ...
````

Reply:
That wouldn't work for `loss=["binary_crossentropy", "categorical_crossentropy"]` or `loss={"out1": "binary_crossentropy", "out2": "categorical_crossentropy"}` which I think are pretty common use cases. So we would still need the recursive part of `_classes_from_strings`. Unless I am missing something?

Reply:
I've updated my comment. `losses` can be used in the `if isinstance(loss, str)` branch in `_classes_from_strings` (which I've renamed to `_build_losses`).

Reply:
So basically we're swapping out the call to `tf.keras.{losses, optimizers, metrics}.get` with a dict lookup, everything else stays the same, right? I feel like I'm missing the big advantage here.

If we're going to give users access to these, as you alluded to in https://github.com/adriangb/scikeras/pull/66#issuecomment-696750092, should we rename `_utils.py` -> `utils.py` and maybe spin off things that really shouldn't be public into another file?

Reply:
> If we're going to give users access to these, as you alluded to in #66 (comment), should we rename _utils.py -> utils.py and maybe spin off things that really shouldn't be public into another file?

I think that's work for a separate PR.

> I feel like I'm missing the big advantage here.

There's not a big advantage. I see a couple smaller advantages:

* The code below encapsulates the logic for processing the return value from `tf.keras.losses.get`. It's not necessary to encode this logic in the processing of user input.
* It makes it easier to have losses/optimizers public (in the future).

``` python
class _Retriever:
    def __init__(self, fn):
        self.fn = fn

    def __getitem__(self, key):
        obj = self.fn(key)
        if isinstance(obj, tf.keras.losses.Loss):
            obj = obj.__class__
        return obj

losses = _Retriever(tf.keras.losses.get)
optimizers = _Retriever(tf.keras.optimizers.get)

# recursive implementation of _build_losses can use losses
```

I think in the future, there should be another class, `shorthands` to get the Keras shorthand from a class/function/instance (so `shorthands[tf.keras.losses.mse] == "mse"` and `shorthands[tf.keras.optimizers.SGD] == "sgd"`). I think `{losses, optimizers, shorthands}` will help users with custom compile functions and processing `compile_kwargs`. I think you have some code that does this.

Reply:
I see. I do agree there are some advantages. If you're okay with it I also think we should leave this for a future PR.

Reply:
This is tracked in #79.

Reply:
> This is tracked in #79.

This test is already removed in this PR. This is an "Outdated" comment thread.